### ISSUE TITLE
Incorrect IncrementalTaskInputs.outOfDate files when using @InputFiles

### ISSUE THREAD

**Issue by scott-pollom** [gradle/gradle#9320]
### Expected Behavior
`IncrementalTaskInputs.outOfDate​(Action<? super InputFileDetails> outOfDateAction)` should execute the action for the correct changed input files, even if the changed input file has the same name and content as another input file. 

### Current Behavior
There are cases when the wrong input file is reported as being `outOfDate`, when the actual changed input file has the same name and content as other input files.

### Steps to Reproduce 
Use the following build.gradle (or use attached [relativeBug.zip](https://github.com/gradle/gradle/files/3129404/relativeBug.zip)):

```

class CustomTask extends DefaultTask {
    @InputFiles
    @PathSensitive(PathSensitivity.NAME_ONLY)
    public FileCollection getInputFiles() {
        return project.files().from('input/a/foo.txt').from('input/b/foo.txt')
    }

    @TaskAction
    def taskAction(IncrementalTaskInputs inputs) {
        println 'outOfDate:'
        inputs.outOfDate { println it.file.absolutePath }
    }
}

task customTask(type: CustomTask) {}

```

In the same directory as this build.gradle file, create the following directory structure with *empty* foo.txt files:

```

input/
  a/
    foo.txt
  b/
    foo.txt

```

Run `$ ./gradlew customTask`. It will correctly list both files as being `outOfDate`.

Add some text to `input/a/foo.txt`, and then run `$ ./gradlew customTask`, and it will correctly list `input/a/foo.txt` as being `outOfDate`

Remove text from `input/a/foo.txt` so it's an empty file again, and then run `$ ./gradlew customTask`, and it will *incorrectly* list `input/b/foo.txt` as being `outOfDate`

### Your Environment
Gradle version: 5.4.1
OS: Linux

**Comment by lptr** [comment]
Thanks @scott-pollom for reporting this. Can you describe an actual use case where this becomes problematic? I'm trying to understand what the right behavior would be.

**Comment by scott-pollom** [comment]
Yes, we've hit this issue with 2 AGP tasks so far, causing the incremental builds to be incorrect in some cases. It's a subtle issue, and we only detected it because the `PackageApplication` task was `UP-TO-DATE` when it shouldn't have been in some of our performance benchmarks. I'm concerned we have other tasks hitting this issue that we're unaware of.

I think the correct behavior would be to report `input/a/foo.txt` as being `outOfDate`, not an unchanged file :)

An example of where we hit this was in AGP's MergeJavaResourceTask, whose input files include `res.jar` files from each of the library modules. If a library module, doesn't have any java resources, the `res.jar` file is empty. If we add a java resource to a library, build, delete the java resource, and build again, we'd get an incorrect build (similar to repro above because content is added to a `res.jar`, and then it becomes empty again).

We've been able to work around this issue in the tasks where we've detected it, but I'm concerned this issue might be affecting other incremental tasks.

**Comment by wolfs** [comment]
@scott-pollom We can do better in detecting non-changed files with the same absolute path. We currently only care about the normalized path for the identity.

Note that, if the path sensitivity is `NAME_ONLY` and the contents of two files are swapped (in your case resources would move from one module to another, making one `res.jar` empty and the other one having the same contents as the original one), then the task will be up to date. Does this also lead to incorrect builds in your case? If it does, then the path sensitivity is probably not `NAME_ONLY`.

**Comment by scott-pollom** [comment]
Thanks, @wolfs.

Yes, the scenario you describe in which the contents of two files are swapped and the task is up-to-date is what I would expect. I hadn't considered that case, and I'll add a test for it, but I expect our build will still be correct in that case because the files are all going to the same place.

My concern is when there is a "net" change in file content and the wrong file is reported as being changed.

**Comment by wolfs** [comment]
This has been fixed via #9354 and will be part of Gradle 5.5.
